docs/WritingTests.asciidoc

```diff
@@ -1065,3 +963,140 @@ the SUSE distribution's serial terminal utility library). However some
+snapshot which can help in this situation.
+
+Depending on the use case, there are two options to help:
+
+* Create and *preserve* snapshots for *every test* module run (+MAKETESTSNAPSHOTS+)
+  - Offers more flexibility as the test can be resumed almost at any point.
+    However disk space requirements are high (expect more than 30GB for one
+    job)
+  - This mode is useful for fixing non-fatal issues in tests and debugging SUT
+    as more than just the snapshot of the last failed module is saved.
+
+* Create a snapshot *after every successful* test module while *always
+  overwriting* the existing snapshot to preserve only the latest (+TESTDEBUG+)
+  - Allows to skip just before the start of the first failed test module,
+    which can be limiting, but preserves disk space in comparison to
+    +MAKETESTSNAPSHOTS+.
+  - This mode is useful for iterative test development
+
+In both modes there is no need to modify tests (i.e. adding +milestone+ test
+flag as the behaviour  is implied). In the later mode every test module is
+also considered +fatal+. This means the job is aborted after the first failed
+test module.
+
+==== Enable snapshots for each module
+
+* Run the worker with --no-cleanup parameter. This will preserve the hard
+ disks after test runs.
+
+* Set +MAKETESTSNAPSHOTS=1+ on a job. This will make openQA save a
+snapshot for every test module run. One way to do that is by cloning an
+existing job and adding the setting:
+
+[source,sh]
+----
+clone_job.pl --from https://openqa.opensuse.org  --host localhost 24 MAKETESTSNAPSHOTS=1
+----
+
+* Create a job again, this time setting the +SKIPTO+ variable to the snapshot you need. Again, +clone_job.pl+ comes handy here:
+
+[source,sh]
+----
+clone_job.pl --from https://openqa.opensuse.org  --host localhost 24 SKIPTO=consoletest-yast2_i
+----
+
+* Use qemu-img snapshot -l something.img to find out what snapshots are in the image. Snapshots are named
+`"test module category"-"test module name"` (e.g. `installation-start_install`).
+
+==== Storing only the last sucessful snapshot
+
+* Run the worker with +--no-cleanup parameter+. This will preserve the hard disks after test runs.
+* Set +TESTDEBUG=1+ on a job. This will make openQA save a snapshot after each
+successful test module run. Snapshots are overwritten. The snapshot is named `lastgood` in all cases.
+
+[source,sh]
+----
+clone_job.pl --from https://openqa.opensuse.org  --host localhost 24 TESTDEBUG=1
+----
+
+* Create a job again, this time setting the +SKIPTO+ variable to the snapshot
+which failed on previous run. Make sure the new job will also have
++TESTDEBUG=1+ set. This can be ensured by the use of the clone_job script on
+the clone source job or specifying the variable explicitly:
+
+[source,sh]
+----
+clone_job.pl --from https://openqa.opensuse.org  --host localhost 24 TESTDEBUG=1 SKIPTO=consoletest-yast2_i
+----
+
+
+=== Tweaking the openQA scheduler for personal development instances
```

Comment:
@asdil12 does this actually work?

Reply:
Please do not document that - we want to make that the default. @mudler just needs a little convincing

Reply:
ah, ok. I would appreciate that :)

@mudler *convince* *convince* *convince* ;)

Reply:
I took it out from this PR and re-pushed

Reply:
Let's try :)

This was never tested on production environment, but it is set on both of our staging instances (it was configured in e122, and e212 is a clone from it). 

```OPENQA_SCHEDULER_SCHEDULE_TICK_MS``` can be safely adjusted as long as congestion control is enabled (which is by default) - scheduler will tune itself during its lifecycle anyway.

```OPENQA_SCHEDULER_WAKEUP_ON_REQUEST``` is a bit of aggressive option, but should be safe anyway, the scheduler will just try to tick as soon as possible when it receives job post/restart requests thru the API, but it does without trying to overload the system, as it will just adjust the clock where  necessary (e.g. if we receive a burst of 40 parallel/subsequent request of scheduler to be restarted, the scheduler will just reset the clock to the default value once).

Aside from that, it's worth to mention that to override systemd services environment variables, you don't need to create a new service, but you can just call ```systemctl edit <service>``` or create the relevant override file like so:

```
systemctl edit openqa-scheduler
cat /etc/systemd/system/openqa-scheduler.service.d/override.conf
[Service]
Environment="OPENQA_SCHEDULER_WAKEUP_ON_REQUEST=1"
```

Reply:
It worked faster than before for me using
```
OPENQA_SCHEDULER_WAKEUP_ON_REQUEST=1 OPENQA_SCHEDULER_SCHEDULE_TICK_MS=1000 ./script/openqa-scheduler
```